lib/delta-table/src/Database/Table/SQL/Stmt.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+{- |
+Copyright: © 2024 Cardano Foundation
+License: Apache-2.0
+
+SQL statements for typed database tables.
+-}
+module Database.Table.SQL.Stmt
+    (
+    -- * SQL statements
+    Stmt
+    , renderStmt
+
+    -- * Specific statements
+    , createTable
+    , selectAll
+    , insertOne
+    , deleteAll
+    ) where
+
+import Prelude
+
+import Data.Text
+    ( Text
+    )
+import Database.Table
+    ( IsTable (getTableName)
+    , getColNames
+    , getTableName
+    )
+import Database.Table.SQL.Column
+    ( SqlType
+    , escapeSqlType
+    )
+import Database.Table.SQL.Table
+    ( IsTableSql
+    , getColumnTypes
+    )
+
+import qualified Data.Text as T
+
+{-------------------------------------------------------------------------------
+    SQL Statements
+-------------------------------------------------------------------------------}
+type TableName = Text
+type ColumnName = Text
+
+-- | A subset of SQL statements.
+data Stmt where
+    CreateTable :: TableName -> [(ColumnName, SqlType)] -> Stmt
+    Select :: [ColumnName] -> TableName -> Where -> Stmt
+    Insert :: TableName -> [ColumnName] -> Stmt
+    Delete :: TableName -> Where -> Stmt
+
+-- | An SQL @WHERE@ clause.
+data Where where
+    All :: Where
+
+{-------------------------------------------------------------------------------
+    Rendering
+-------------------------------------------------------------------------------}
+-- | Render an statement as SQL source code.
+renderStmt :: Stmt -> Text
+renderStmt (CreateTable table cols) =
+    "CREATE TABLE IF NOT EXISTS "
+        <> renderName table
+        <> " " <> renderTuple (map renderCol cols)
+        <> ";"
+  where
+    renderCol (col, typ)= renderName col <> " " <> escapeSqlType typ
+renderStmt (Insert table cols) =
+    "INSERT INTO "
+        <> renderName table
+        <> " " <> renderTuple (map renderName cols)
+        <> " VALUES " <> renderTuple ("?" <$ cols)
+        <> ";"
+renderStmt (Select cols table All) =
+    "SELECT " <> T.intercalate "," (map renderName cols)
+        <> " FROM " <> renderName table
+        <> ";"
+renderStmt (Delete table All) =
+    "DELETE FROM " <> renderName table
+
+-- | Escape a column or table name.
+renderName :: Text -> Text
+renderName s = "\"" <> s <> "\""
+
+-- | Render a tuple
+renderTuple :: [Text] -> Text
+renderTuple xs = "(" <> T.intercalate ", " xs <> ")"
+
+{-------------------------------------------------------------------------------
+    SQL queries
+-------------------------------------------------------------------------------}
+
+-- | Create a database table.
+createTable :: IsTableSql t => proxy t -> Stmt
+createTable proxy =
+    CreateTable
+        (getTableName proxy)
+        (zip (getColNames proxy) (getColumnTypes proxy))
```

Comment:
> suspect you did this to keep Database.Table.Schema separate from any notion of SQL/Sqlite though.

Indeed.

I also would have liked the module `Database.Table.SQL.Column` to not depend on specific types from `sqlite-simple`, because the SQL language is, in principle, independent of any particular backend such as SQLite. But that would have meant duplicating the types, and I left it at this.